session/sql_migration_test.go

```diff
@@ -0,0 +1,738 @@
+			// randomized if they exist or not.
+			macCaveat.Id = randomBytes(rand.Intn(10) + 1)
+
+			if rand.Intn(2) == 0 {
+				macCaveat.VerificationId =
+					randomBytes(rand.Intn(32) + 1)
+			}
+
+			if rand.Intn(2) == 0 {
+				macCaveat.Location =
+					randomString(rand.Intn(10) + 1)
+			}
+
+			macCaveats = append(macCaveats, macCaveat)
+		}
+	} else {
+		macCaveats = nil
+	}
+
+	// We can't do both nil caveats and nil perms, so if we have nil
+	// caveats, we set perms to a value.
+	if rand.Intn(2) == 0 || macCaveats == nil {
+		for i := 0; i < loopLen; i++ {
+			var macPerm bakery.Op
+
+			macPerm.Action = randomString(rand.Intn(10) + 1)
+			macPerm.Entity = randomString(rand.Intn(10) + 1)
+
+			macPerms = append(macPerms, macPerm)
+		}
+	} else {
+		macPerms = nil
+	}
+
+	return WithMacaroonRecipe(macCaveats, macPerms)
+}
+
+// randomAccountOptions creates a random account with a random balance and
+// expiry time, that's linked in the returned options. The returned options also
+// returns the macaroon recipe with the account caveat.
+func randomAccountOptions(ctx context.Context, t *testing.T,
+	acctStore accounts.Store) []Option {
+
+	balance := lnwire.MilliSatoshi(rand.Int63())
+
+	// randomize expiry from 10 to 10,000 minutes
+	expiry := time.Now().Add(
+		time.Minute * time.Duration(rand.Intn(10000-10)+10),
+	)
+
+	// As the store has a unique constraint for inserting labels, we suffix
+	// it with a sufficiently large random number avoid collisions.
+	label := fmt.Sprintf("account:%d", rand.Int63())
+
+	// Create an account with balance
+	acct, err := acctStore.NewAccount(
+		ctx, balance, expiry, label,
+	)
+	require.NoError(t, err)
+	require.False(t, acct.HasExpired())
+
+	// For now, we manually add the account caveat
+	// for bbolt compatibility.
+	accountCaveat := checkers.Condition(
+		macaroons.CondLndCustom,
+		fmt.Sprintf("%s %x",
+			accounts.CondAccount,
+			acct.ID[:],
+		),
+	)
+
+	sessCaveats := []macaroon.Caveat{}
+	sessCaveats = append(
+		sessCaveats,
+		macaroon.Caveat{
+			Id: []byte(accountCaveat),
+		},
+	)
+
+	opts := []Option{
+		WithAccount(acct.ID),
+		WithMacaroonRecipe(sessCaveats, nil),
+	}
+
+	return opts
+}
+
+// randomBytes generates a random byte array of the passed length n.
+func randomBytes(n int) []byte {
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = byte(rand.Intn(256)) // Random int between 0-255, then cast to byte
+	}
+	return b
+}
+
+// randomString generates a random string of the passed length n.
+func randomString(n int) string {
+	letterBytes := "abcdefghijklmnopqrstuvwxyz"
+
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = letterBytes[rand.Intn(len(letterBytes))]
+	}
+	return string(b)
+}
+
+// shiftStateUnsafe updates the state of the session with the given ID to the
+// "dest" state, without checking if the state transition is legal.
+//
+// NOTE: this function should only be used for testing purposes.
+func (db *BoltStore) shiftStateUnsafe(_ context.Context, id ID,
```

Comment:
Not fully sure if we add test functions like this one that's a `struct` member to the `_test` file, or if we add it in the actual struct file?

Reply:
right, not too great, you could change to `func shiftStateUnsafe(_ context.Context, db *BoltStore, id ID, dest State) error {`